### ISSUE TITLE
Collision detection problem

### ISSUE THREAD

**Issue by JianchengYang** [flexible-collision-library/fcl#481]
How to do collision detection between a line segment and a cube in space?I have checked the introduction, but the above can only constitute two cubes for detection, because I only know two point coordinates and cannot calculate the attitude matrix, so I have a problem.

Reference code:
`// Given two objects o1 and o2
CollisionObjectf* o1 = ...
CollisionObjectf* o2 = ...
// set the collision request structure, here we just use the default setting
CollisionRequest request;
// result will be returned via the collision result structure
CollisionResult result;
// perform collision test
collide(o1, o2, request, result);`

**Comment by SeanCurtis-TRI** [comment]
There is no direct support for a line segment, per se. But you have several options:

1. Represent the line segment as a long skinny geometry (cylinder, box, etc.) The thicker the geometry the greater the error, but you should be able to create a box/cylinder/capsule that essentially has zero volume and get a reasonable solution. If you go this route, I'd recommend using a box instead of a cylinder or capsule. There is an explicit primitive collision function between boxes. Box-cylinder and box-capsule defaults to using GJK which is going to produce inherently inferior answers.
    - If you do this, you'll need to orient the "line". You can create a basis for it by orienting the box's x-axis along the direction of the line segment. Then you can use `generateCoordinateSystem` in [`math/geometry.h`](https://github.com/flexible-collision-library/fcl/blob/master/include/fcl/math/geometry.h) to construct a basis from it. This API will change a bit in the near future; the pr #467 is currently fixing the underlying logic, but it should still suffice for your needs. Or, if it doesn't, you can steal the code from that PR and use it directly.
2. You *might* be able to use the continuous collision detection. I haven't worked with FCL's continuous collision detection so this suggestion may not be correct. But, if it works, you'd use a box with a fixed pose over the interval, and a point that moves from one endpoint of your segment to the other end.

**Comment by JianchengYang** [comment]
> There is no direct support for a line segment, per se. But you have several options:
> 
> 1. Represent the line segment as a long skinny geometry (cylinder, box, etc.) The thicker the geometry the greater the error, but you should be able to create a box/cylinder/capsule that essentially has zero volume and get a reasonable solution. If you go this route, I'd recommend using a box instead of a cylinder or capsule. There is an explicit primitive collision function between boxes. Box-cylinder and box-capsule defaults to using GJK which is going to produce inherently inferior answers.
>    
>    * If you do this, you'll need to orient the "line". You can create a basis for it by orienting the box's x-axis along the direction of the line segment. Then you can use `generateCoordinateSystem` in [`math/geometry.h`](https://github.com/flexible-collision-library/fcl/blob/master/include/fcl/math/geometry.h) to construct a basis from it. This API will change a bit in the near future; the pr #467 is currently fixing the underlying logic, but it should still suffice for your needs. Or, if it doesn't, you can steal the code from that PR and use it directly.
> 2. You _might_ be able to use the continuous collision detection. I haven't worked with FCL's continuous collision detection so this suggestion may not be correct. But, if it works, you'd use a box with a fixed pose over the interval, and a point that moves from one endpoint of your segment to the other end.

Thank you for your reply!
My problem now is how to determine the transformation matrix if two spatial points constitute a cube. I did not find relevant materials in the Math library you mentioned, could you please explain it in detail?Is there a way that continuous collision detection can move?

**Comment by SeanCurtis-TRI** [comment]
I didn't try compiling this and there's the very real possibility of some slight error that prevents this from compiling. However, the basic shape of things is as follows:

```c++
/** Construct a basis B for a box aligned with the given line segment.

The x-axis of the basis is aligned with the line segment defined by
points X0 and X1 (both measured and expressed in the same frame F).
Specifically, the x-axis will point in the same direction as the vector *from*
X0 *to* to X1.

@param p_FX0   One of the segment end points (X0) measured and
expressed in frame F.
@param p_FX1   The other segment end point (X1) measured and
expressed in frame F.
@retval R_FB The relative orientation between frame F and the box's
frame B.
@pre |X0 - X1| > 1e-5 (to make sure that garbage values aren't introduced). */
template <typename S>
Matrix3<S> CalcBasisAlongSegment(const Vector3<S>& p_FX0, const Vector3<S>& p_FX1) {
  const Vector3<S> p_X0X1_F = p_FX1 - p_FX0;
  assert(p_X0X1_F.squaredNorm() >= 1e-5 * 1e-5);
  Matrix3<S> R_FB;
  R_FB.col(0) = p_X0X1_F.normalized();
  generateCoordinateSystem(R_FB);
  return R_FB;
}
```

**Comment by JianchengYang** [comment]
> I didn't try compiling this and there's the very real possibility of some slight error that prevents this from compiling. However, the basic shape of things is as follows:
> 
> ```c++
> /** Construct a basis B for a box aligned with the given line segment.
> 
> The x-axis of the basis is aligned with the line segment defined by
> points X0 and X1 (both measured and expressed in the same frame F).
> Specifically, the x-axis will point in the same direction as the vector *from*
> X0 *to* to X1.
> 
> @param p_FX0   One of the segment end points (X0) measured and
> expressed in frame F.
> @param p_FX1   The other segment end point (X1) measured and
> expressed in frame F.
> @retval R_FB The relative orientation between frame F and the box's
> frame B.
> @pre |X0 - X1| > 1e-5 (to make sure that garbage values aren't introduced). */
> template <typename S>
> Matrix3<S> CalcBasisAlongSegment(const Vector3<S>& p_FX0, const Vector3<S>& p_FX1) {
>   const Vector3<S> p_X0X1_F = p_FX1 - p_FX0;
>   assert(p_X0X1_F.squaredNorm() >= 1e-5 * 1e-5);
>   Matrix3<S> R_FB;
>   R_FB.col(0) = p_X0X1_F.normalized();
>   generateCoordinateSystem(R_FB);
>   return R_FB;
> }
> ```

I've done it. Thank you

**Comment by SeanCurtis-TRI** [comment]
You should know that PR #467 is about to change the interface of `generateCoordinateSystem()`.
